packages/zero-solid/src/use-zero.ts

```diff
@@ -0,0 +1,77 @@
+import {
+  batch,
+  createContext,
+  createMemo,
+  onCleanup,
+  useContext,
+  type JSX,
+} from 'solid-js';
+import {
+  Zero,
+  type CustomMutatorDefs,
+  type Schema,
+  type ZeroOptions,
+} from '../../zero/src/zero.ts';
+
+// eslint-disable-next-line @typescript-eslint/naming-convention, @typescript-eslint/no-explicit-any
+const ZeroContext = createContext<(() => Zero<any, any>) | undefined>(
+  undefined,
+);
+
+export function createZero<S extends Schema, MD extends CustomMutatorDefs<S>>(
+  options: ZeroOptions<S, MD>,
+): Zero<S, MD> {
+  const opts = {
+    ...options,
+    batchViewUpdates: batch,
+  };
+  return new Zero(opts);
+}
+
+export function useZero<
+  S extends Schema,
+  MD extends CustomMutatorDefs<S> | undefined = undefined,
+>(): () => Zero<S, MD> | undefined {
+  const zero = useContext(ZeroContext);
+
+  // TODO: uncomment when we require ZeroProvider in a future release.
+  // if (zero === undefined) {
+  //   throw new Error('useZero must be used within a ZeroProvider');
+  // }
+  return zero ?? (() => undefined);
+}
+
+export function createUseZero<
+  S extends Schema,
+  MD extends CustomMutatorDefs<S> | undefined = undefined,
+>() {
+  return () => useZero<S, MD>();
+}
+
+export function ZeroProvider<
+  S extends Schema,
+  MD extends CustomMutatorDefs<S> | undefined = undefined,
+>(props: {
+  children: JSX.Element;
+  zeroSignal: () => ZeroOptions<S, MD> | {zero: Zero<S, MD>};
+}) {
+  const zero = createMemo(() => {
+    const z = props.zeroSignal();
+    if ('zero' in z) {
+      return z.zero;
+    }
+    const createdZero = new Zero({
+      ...z,
+      batchViewUpdates: batch,
+    });
+    onCleanup(() => createdZero.close());
+    return createdZero;
+  });
+
+  return ZeroContext.Provider({
+    value: zero,
+    get children() {
```

Comment:
so it seems this getter for children was the secret sauce... but why?

Why does the following not work
```
return ZeroContext.Provider({
  value: zero,
  children: props.children
  });
  
```

Reply:
Should value also be a getter?  There is still something fundamental about solidjs that i dont grok.

Reply:
Ah you can't use JSX here

In that case the getter is correct.
value would also need to be a getter.

In solid when you access something from the props like `props.value` inside a reactive scope, it needs to behave similar to accessing a signal `value()` so that it can be tracked as a dependency. So the JSX transform changes `<A b={c()} />` to `A({ get b() { return c{} } })`.

This makes sure that every property passed through the component tree can be tracked independently.